apps/emqx_durable_storage/src/emqx_ds_lts.erl

```diff
@@ -0,0 +1,509 @@
+    Varying = case trie_next_(Trie, State, Tok) of
+                  {NChildren, _, _DiscardState} when is_integer(NChildren), NChildren > Threshold ->
+                      {_, NextState} = trie_insert(Trie, State, ?PLUS),
+                      [Tok|Varying0];
+                  {_, false, NextState} ->
+                      Varying0;
+                  {_, true, NextState} ->
+                      [Tok|Varying0]
+              end,
+    do_topic_key(Trie, ThresholdFun, Depth + 1, NextState, Rest, Varying).
+
+-spec trie_next_(trie(), state(), binary() | ?EOT) -> {New, Wildcard, state()}
+              when New :: false | non_neg_integer(),
+                   Wildcard :: boolean().
+trie_next_(Trie, State, Token) ->
+    case trie_next(Trie, State, Token) of
+        {Wildcard, NextState} ->
+            {false, Wildcard, NextState};
+        undefined ->
+            {Updated, NextState} = trie_insert(Trie, State, Token),
+            {Updated, false, NextState}
+    end.
+
+%% @doc Return all edges emanating from a node:
+%% erlfmt-ignore
+-spec emanating(trie(), state(), edge()) -> [{edge(), state()}].
+emanating(#trie{trie = Tab}, State, ?PLUS) ->
+    ets:select(Tab, ets:fun2ms(
+                      fun(#trans{key = {S, Edge}, next = Next}) when S == State ->
+                              {Edge, Next}
+                      end));
+emanating(#trie{trie = Tab}, State, ?EOT) ->
+    case ets:lookup(Tab, {State, ?EOT}) of
+        [#trans{next = Next}] -> [{?EOT, Next}];
+        []                    -> []
+    end;
+emanating(#trie{trie = Tab}, State, Bin) when is_binary(Bin) ->
+    [{Edge, Next} || #trans{key = {_, Edge}, next = Next} <-
+                         ets:lookup(Tab, {State, ?PLUS}) ++
+                         ets:lookup(Tab, {State, Bin})].
+
+%%================================================================================
+%% Tests
+%%================================================================================
+
+-ifdef(TEST).
+
+trie_basic_test() ->
+    T = trie_create(),
+    ?assertMatch(undefined, trie_next(T, ?PREFIX, <<"foo">>)),
+    {1, S1} = trie_insert(T, ?PREFIX, <<"foo">>),
+    ?assertMatch({false, S1}, trie_insert(T, ?PREFIX, <<"foo">>)),
+    ?assertMatch({false, S1}, trie_next(T, ?PREFIX, <<"foo">>)),
+
+    ?assertMatch(undefined, trie_next(T, ?PREFIX, <<"bar">>)),
+    {2, S2} = trie_insert(T, ?PREFIX, <<"bar">>),
+    ?assertMatch({false, S2}, trie_insert(T, ?PREFIX, <<"bar">>)),
+
+    ?assertMatch(undefined, trie_next(T, S1, <<"foo">>)),
+    ?assertMatch(undefined, trie_next(T, S1, <<"bar">>)),
+    {1, S11} = trie_insert(T, S1, <<"foo">>),
+    {2, S12} = trie_insert(T, S1, <<"bar">>),
+    ?assertMatch({false, S11}, trie_next(T, S1, <<"foo">>)),
+    ?assertMatch({false, S12}, trie_next(T, S1, <<"bar">>)),
+
+    ?assertMatch(undefined, trie_next(T, S11, <<"bar">>)),
+    {1, S111} = trie_insert(T, S11, <<"bar">>),
+    ?assertMatch({false, S111}, trie_next(T, S11, <<"bar">>)).
```

Comment:
Tiny suggestion: assert that all returned states S1, S2, ..., are distinct.

Reply:
This is done automatically by trie_next